src/NuGet.Core/NuGet.Packaging/Rules/DependeciesGroupsForEachTFMRule.cs

```diff
@@ -0,0 +1,191 @@
+                    {
+                        if (!tfmsFromNuspec.Contains(fileTFM))
+                        {
+                            _noExactMatchesFromFile.Add(fileTFM);
+                        }
+
+                        if (!tfmsFromFiles.Contains(nuspecTFM))
+                        {
+                            _noExactMatchesFromNuspec.Add(nuspecTFM);
+
+                            if (isCompatible(fileTFM, nuspecTFM))
+                            {
+                                _compatNotExactMatches.Add(fileTFM);
+                            }
+                        }
+                    }
+                }
+
+                _noExactMatchesFromFile.RemoveWhere(IsCompat);
+                _noExactMatchesFromNuspec.RemoveWhere(IsCompat);
+                (string noExactMatchString, string compatMatchString) =
+                    GenerateWarningString(_noExactMatchesFromFile, _noExactMatchesFromNuspec, _compatNotExactMatches);
+
+                var issues = new List<PackagingLogMessage>();
+
+                if (_noExactMatchesFromFile.Count != 0 || _noExactMatchesFromNuspec.Count != 0)
+                {
+                    issues.Add(PackagingLogMessage.CreateWarning(string.Format(MessageFormat,noExactMatchString),
+                        NuGetLogCode.NU5128));
+                }
+
+                if (_compatNotExactMatches.Count != 0)
+                {
+                    issues.Add(PackagingLogMessage.CreateWarning(string.Format(SecondWarning, compatMatchString),
+                        NuGetLogCode.NU5130));
+                }
+
+                if(issues.Count != 0)
+                {
+                    return issues;
+                }
+            }
+            else
+            if((tfmsFromFiles.Count != 0 && tfmsFromNuspec.Count == 0) || (tfmsFromFiles.Count == 0 && tfmsFromNuspec.Count != 0))
+            {
+                if(tfmsFromFiles.Count != 0)
+                {
+                    _noExactMatchesFromFile.AddRange(tfmsFromFiles);
+                }
+
+                if(tfmsFromNuspec.Count != 0)
+                {
+                    _noExactMatchesFromNuspec.AddRange(tfmsFromNuspec);
+                }
+
+                (string noExactMatchString, string compatMatchString) =
+                    GenerateWarningString(_noExactMatchesFromFile, _noExactMatchesFromNuspec, _compatNotExactMatches);
+
+                var issues = new List<PackagingLogMessage>();
+                issues.Add(PackagingLogMessage.CreateWarning(string.Format(MessageFormat, noExactMatchString),
+                        NuGetLogCode.NU5128));
+                return issues;
+            }
+
+            return Array.Empty<PackagingLogMessage>();
+        }
+
+        private bool IsCompat(NuGetFramework tfm)
+        {
+            return _compatNotExactMatches.Contains(tfm);
+        }
+
+        private static IEnumerable<ContentItemGroup> GetContentForPattern(ContentItemCollection collection, PatternSet pattern)
```

Comment:
suggestion: We have many rules that reuse the same methods. They can refactored to a util :)